### PR TITLE
Python3 compatibility updates

### PR DESCRIPTION
**Pull request by petiepooo**
Passes test suite when run under python 2.7.15rc1 and 3.6.7.  Has not yet been tested under live system.

### CODE REVIEW

**Comment by petiepooo** [comment]
I can verify this works on python 2.7.3.  I have not tested on 3.x at all; please validate before merging.

**Comment by petiepooo** [comment]
Testsuite also passes on python versions 2.7.12 and 3.5.2.  I'm still unable to test in pure python3 environment; sorry...

**Comment by kzw** [comment]
I tested this and it did not work.  I look in the openvpn log and do not find anything.  If someone know how to capture, it will help.  Otherwise I can dig deeper.

Code from master branch works with python2.  I simply replace two files (duo_openvpn.py and https_wrapper.py) in this PR with what are on the working machine.  Also changed `/usr/bin/python` symlink to point to `python3.5`.  I then downloaded the tarball from the PR at ``` wget https://github.com/duosecurity/duo_openvpn/tarball/db9e94aa8a1771971f6213e7437774a73e2f70b4 -o duo.tar.gz``` and then run `make && make install` which is the same command used in our unattended install process.  This also fail to work.

Tested with ubuntu 16.04.  python 3.5.2.

**Comment by petiepooo** [comment]
@kzw i'm sorry to hear that.  May I please ask you to provide a few more details?
You are right to edit the file to explicitly use python3 by changing the first line.  Does it work if you set it to use python2?
When you went to connect, did it fail within a few seconds, or did it wait for 30 or more seconds before failing?  That may give an indication on whether it was able to launch the interpreter from the compiled plugin.
Please also check your syslog (`/var/log/syslog`) for strings containing `Duo OpenVPN` from around that time.  If it was able to launch the interpreter, those entries should give an indication of why it failed.
Also, only the latest version doesn't hard-code the interpreter in the compiled plugin (see #18).  I'd recommend you only test with your freshly compiled plugin.  You'd have to restart openvpn afterwards, of course, and you can verify the path of what was loaded by running `lsof -p [pid]` using the pid of your openvpn process and looking for the `duo_openvpn.so` line.
If it's taking 30seconds or more, you can also view the invoked python command running `pstree -ap $(pgrep -o openvpn)` right after attempting to connect to verify that it's running python3.5 explicitly.  You can use the pid of the python process to view the environment it was passed to by running `sudo cat /proc/[pid]/environ |tr '\x0' '\n'`
I appreciate your time in testing this.

**Comment by petiepooo** [comment]
You can also try adding the line `log('Python version %s' % sys.version)` around line#21 (just after log is defined) and checking syslog after running it.. just to make sure you're in the right interpreter.

**Comment by petiepooo** [comment]
Never mind.. I have a way to test now, and I'm seeing an immediate failure with a 401-unauthorized.  Probably due to the way strings are handled in py3 and how the hash is formed.  I'll try to have an additional patch in the PR once I'm able to figure that out..

**Comment by kzw** [comment]
```May 22 14:56:52 test-bakery-instance-zz /duo_openvpn.py: Duo OpenVPN: Received 401 Unauthorized
May 22 14:56:52 test-bakery-instance-zz /duo_openvpn.py: Duo OpenVPN: writing failure code to /tmp/openvpn_acf_21ccfc1676b670683d17a216b2005196.tmp```

I can confirm this.  We should have a test for this failure.

**Comment by petiepooo** [comment]
ok, with these last three patches, I have this working on both 2.7.3 and 3.2.3, tested to the point of actually being able to authenticate and connect.

FWIW, after a422d03, the existing testsuit failed when running under python 3.6.7; I just wasn't able to run tests on my openvpn system (too old to be able to install python3-mox3; yeah, I know...).  f0242fb fixed it so the testsuite passes again.  It seems to stem from the differences between python2's httplib (data field is str) and python3's http.client (data field is bytes), and possibly a difference when attempting to decode a string that's already decoded between python v3.2 and v3.6.

c4de8e6 was an oversight during my initial work in ef341a6.  I don't know that we need a separate testcase just for that.

Anyway, please give it a shot now and see if it's working for you.

**Comment by petiepooo** [comment]
ok, with these last three patches, I have this working on both 2.7.3 and 3.2.3, tested to the point of actually being able to authenticate and connect.

FWIW, after a422d03, the existing testsuit failed when running under python 3.6.7; I just wasn't able to run tests on my openvpn system (too old to be able to install python3-mox3; yeah, I know...).  f0242fb fixed it so the testsuite passes again.  It seems to stem from the differences between python2's httplib (data field is str) and python3's http.client (data field is bytes), and possibly a difference when attempting to decode a string that's already decoded between python v3.2 and v3.6.

c4de8e6 was an oversight during my initial work in ef341a6.  I've enhanced an existing test to check for that now...

Anyway, please give it a shot now and see if it's working for you.

**Comment by kzw** [comment]
It works now.  Thanks for these fixes 💯

**Comment by petiepooo** [comment]
Great to hear!  And you're welcome. I'll be going through the same process soon enough, so this will help me out too.

I'll work with Duo support to let them know it should be ready to merge.

**Comment by vbscott** [comment]
Hi, thank you for the pull request.  We put this request on our backlog and will take a look at it.  One of the first things we noticed is that you do not use a Python 2 and 3 compatibility library.  Would you consider updating your pull request to use six?

**Comment by xdesai** [comment]
**Summary:**
This PR was really close I just made a few edits. Thanks @petiepooo for your work on this.
Updated the PR to use six for the py2->py3 compat layer.
Made a couple edits to the PR around encoding, and making it as similar to duo_client_python as possible.

**Test Plan:**
Automated tests pass.
Manually tested with python2 and python3
push, sms, phone call

File: test_duo_openvpn.py [hunk 1]
```diff
@@ -80,6 +80,8 @@ def assert_auth(self, environ, expected_control,
 
             control.seek(0, os.SEEK_SET)
             output = control.read()
+            if not isinstance(output, six.text_type):
+                output = output.decode('ascii')
```
Comment (mbish) [inline test_duo_openvpn.py]: Can we stick to `utf8` here instead?

**Pull request merged**